Python/parking_lot.c

```diff
@@ -194,14 +194,17 @@ _PySemaphore_Wait(_PySemaphore *sema, PyTime_t timeout, int detach)
     PyThreadState *tstate = NULL;
     if (detach) {
         tstate = _PyThreadState_GET();
-        if (tstate) {
-            PyEval_ReleaseThread(tstate);
+        // Only detach if we are attached
+        if (tstate && tstate->state != _Py_THREAD_ATTACHED) {
+            tstate = NULL;
         }
     }
 
+    if (tstate) {
+        PyEval_ReleaseThread(tstate);
+    }
```

Comment:
Moving the check-and-release out of the `detach` branch implies some degree of applicability outside that one branch.  Do you have some future alternate branches in mind?  If not, it would be a little clearer to move that back into the branch:

```suggestion
        if (tstate && tstate->state != _Py_THREAD_ATTACHED) {
            tstate = NULL;
        }
        if (tstate) {
            PyEval_ReleaseThread(tstate);
        }
    }
```

or even:

```suggestion
        if (tstate) {
            if (tstate->state != _Py_THREAD_ATTACHED) {
                tstate = NULL;
            }
            else {
                PyEval_ReleaseThread(tstate);
            }
        }
    }
```

Reply:
I moved it back in the branch, but put the `PyEval_ReleaseThread()` case first